### ISSUE TITLE
RFC: productionize deterministic Qwen3.5 MoE routing for MTP

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#24]
## Background

While enabling Qwen3.5 MXFP4 MTP, we expected greedy speculative decoding to be target-equivalent: the draft model proposes tokens, the target model verifies them, and any mismatch should fall back to the target token. In practice, the first AIME25 MTP eval exposed a score drop in one mode, so we debugged it as a target-equivalence/stability problem rather than as an eval-only issue.

The key finding was that the drift was not purely an MTP cache bug. A smaller TP=1 Qwen3.5 35B repro showed early text divergence between non-MTP and MTP on the same prompt. We then tried forcing MTP accept length to 1, which still followed the MTP trajectory. Direct kernel checks showed the obvious primitives were not the culprit: MHA multi-query decode matched repeated q=1 decode, GDN multi-token verify matched repeated decode when using the real float32 recurrent state, and Triton MXFP4 precomputed MoE row 0 was batch-invariant in isolation.

The stronger control was four identical non-MTP requests in the same batch. Those identical requests produced different text trajectories. That means MTP is exposing a broader batch-shape/routing determinism issue: MTP changes the target verification shape from roughly `batch_size` rows to `batch_size * draft_width` rows, and this can perturb which expert ids are chosen when router scores tie exactly or nearly exactly.

## Current mitigation

Commit `20dba7f` makes native MoE top-k tie handling deterministic by replacing `torch.topk(..., sorted=False)` in the torch-native routing helper with a stable descending sort. Exact equal scores now break ties by expert order, for both normal routing and bias-corrected routing.

That commit also disables SMG health checks under `ts serve` by default. Previously the gateway health probe could be scheduled as real `HEALTH_CHECK_*` generation work during long evals, which changes batch composition while measuring deterministic greedy behavior.

With the current branch and MTP enabled, the local TP=4 Qwen3.5 397B MXFP4 AIME25 limit-8 run scored `1.0` with `enable_thinking=false`, Avg TPOT around `9.24ms`, and throughput around `100.74 tok/s`.

## Why the current fix is not enough long-term

The current stable sort is correct and simple, but it is not the final production-quality shape:

- `torch.sort(..., stable=True)` over all experts is more work than top-k selection. It is fine for correctness and a small top-k helper, but it can become expensive as expert count grows.
- It only handles the torch-native routing helper. Other routing paths may still have their own tie behavior: grouped top-k, CUDA/third-party routing, FlashInfer-style routing, or model-specific routing kernels.
- Exact tie stability is necessary, but near-tie stability may still matter. If logits differ by a few ulps because batch geometry or collective order changes, the engine can still choose different experts.
- We do not yet have an end-to-end deterministic routing test that compares identical requests across batch sizes, or MTP target verify versus non-MTP decode.

## Goal

Make Qwen3.5 MoE routing deterministic across batch shape, request order, and MTP target-verify width, without paying full-sort overhead in production.

The production-ready contract should be:

1. Identical router scores choose identical expert ids independent of batch size and request position.
2. Equal-score ties break by a documented rule, ideally ascending expert id after descending score.
3. The top-k output is stable across the routing implementations used by Qwen3.5 MXFP4 on AMD.
4. The implementation remains O(E log K) or similar top-k complexity, not full O(E log E), for production routing.
5. MTP greedy eval remains target-equivalent enough that any residual accuracy variation is model/numeric variance, not deterministic routing drift.

### 1. Build a minimal deterministic repro test

Add a runtime unit test for `torch_native_fused_topk` with exact ties. This already exists in the current mitigation, but should be expanded into a shared routing contract test.

Test cases:

- All experts exactly tied: `router_logits = [0, 0, 0, 0]`, `top_k=2`, expected ids `[0, 1]`.
- Partial tie at the top: `router_logits = [2, 1, 2, 1]`, expected ids `[0, 2]`.
- Bias-corrected routing: scores used for choice include correction bias, but weights are gathered from original softmax scores. Equal biased scores should also break by expert id.

### 2. Add batch-shape invariance tests

Create a CPU/GPU test that routes the same row in two contexts:

- Once as shape `[1, num_experts]`.
- Once as row 0 of shape `[N, num_experts]`, where the other rows contain arbitrary logits.

Assert row 0 returns identical `topk_ids` and `topk_weights`.

This should run for:

- torch-native top-k path.
- grouped top-k path if Qwen3.5 or related models can use it.
- bias-corrected path if available on the platform.

### 3. Replace full stable sort with deterministic top-k

The current stable sort is easy to reason about but too expensive to leave as the only production path if expert count grows. A better implementation is to make the key lexicographic:

```text
primary: score descending
secondary: expert_id ascending
```

Options:

- For small expert counts, keep `stable=True` sort because the overhead is probably negligible compared with model work and it is easy to validate.
- For larger expert counts, implement a deterministic top-k kernel that compares `(score, -expert_id)` or `(score, tie_break)` in one pass.
- For torch-only fallback, use a tiny expert-id epsilon only if it is provably safe. Be careful: adding epsilon to floating scores can change legitimate near-ties. Prefer lexicographic comparison over numerical perturbation.

A production Triton/CUDA/Cute/Gluon routing kernel should explicitly encode the tie rule instead of relying on library `topk` behavior.

### 4. Audit all routing paths

Search for every place we use `torch.topk`, `topk`, `routing_flash`, grouped routing, FlashInfer routing, or model-specific top-k.

For each path, answer:

- Is it used by Qwen3.5 MXFP4 on AMD?
- Is it used by MTP target verify or only by the draft model?
- What is its tie behavior?
- Does it produce sorted ids or only sorted scores?
- Does expert-id tie order matter downstream?

Then update each relevant path to share the same contract.

### 5. Add an end-to-end batch-shape smoke test

Use a small model or a synthetic MoE block if full Qwen3.5 is too expensive for CI.

Desired test shape:

- Send the same prompt once with batch size 1.
- Send multiple identical prompts concurrently.
- Compare the first N generated token ids for request 0.

This does not need to run in every PR if it is expensive, but it should exist as a manual or nightly regression test for Qwen3.5/MTP.

### 6. Add MTP equivalence diagnostics

For debugging only, add an opt-in trace mode that logs or dumps:

- candidate draft tokens
- target predicted tokens for verify rows
- accepted lengths
- per-layer first divergence probes, ideally disabled by default

This should not be a production feature. The goal is to avoid spending days rediscovering whether a future MTP drift comes from target logits, verifier acceptance, draft cache, or routing.

## Acceptance criteria

- Unit tests prove exact top-k ties resolve by expert id for normal and bias-corrected routing.
- Batch-shape invariance test passes for the Qwen3.5 routing path used on AMD.
- No SMG `HEALTH_CHECK_*` generation requests appear during `ts serve` eval runs after startup readiness.
- TP=4 Qwen3.5 397B MXFP4 AIME25 limit-8 MTP eval with `enable_thinking=false` meets or exceeds the existing `0.875` threshold.
- Perf regression from deterministic tie-breaking is measured and either negligible or replaced with a production top-k kernel.

## Notes from current local validation

Current MTP eval command shape:

```text
ts serve --model /data/models/amd-Qwen3.5-397B-A17B-MXFP4 \
  --tensor-parallel-size 4 \
  --max-model-len 80000 \
  --max-num-seqs 16 \
  --quantization mxfp4 \
  --speculative-algorithm MTP \
  --speculative-num-steps 3 \
  --speculative-num-draft-tokens 4 \
  --speculative-draft-model-quantization unquant \
  --sampling-backend greedy \
  --disable-kvstore --kvstore-ratio 0.0
```

Eval:

```text
evalscope eval \
  --datasets aime25 \
  --limit 8 \
  --eval-batch-size 8 \
  --stream \
  --generation-config '{"do_sample":false,"temperature":0.0,"max_tokens":16384,"extra_body":{"chat_template_kwargs":{"enable_thinking":false}}}'
```

Latest observed local result after the stable-routing mitigation:

- score: `1.0`
- wall time: `164s`
- Avg TTFT: `1460.33ms`
- Avg TPOT: `9.24ms`
- Avg throughput: `100.74 tok/s`
